test/dlc_tests/benchmark/sagemaker/tensorflow/training/test_performance_tensorflow_sm_training.py

```diff
@@ -0,0 +1,56 @@
+import os
+import random
+import re
+import time
+
+import pytest
+
+from invoke.context import Context
+
+from test.test_utils import BENCHMARK_RESULTS_S3_BUCKET
+
+
+@pytest.mark.parametrize("num_nodes", [1, 4], indirect=True)
+def test_tensorflow_sagemaker_training_performance(tensorflow_training, num_nodes, region):
+
+    # This sleep has been inserted because all the parametrized training jobs are automatically created
+    # by SageMaker with the same name, due to being started around the same time, and with the same image uri.
+    time.sleep(random.Random(x=f"{tensorflow_training}{num_nodes}").random() * 60)
```

Comment:
Can you please explain the functionality of this line?

Reply:
We can have TensorFlow Training images for CPU/GPU, Py2/Py3, TF-1 and TF-2, all of which can be performance tested using the same test function.

It is unlikely that both TF-1 and TF-2 images will be tested in the same process, but it is highly likely that 2-4 images (CPU/GPU-Py2/Py3/Py37) will be tested as a part of the same CodeBuild job.

When these threads are started within moments of one-another, the SageMaker Training Job that they create for the benchmark test will also be created around the same time. Each of these jobs has a "unique" name, that is automatically generated by SageMaker based on the time at which the test was started. However, this unique name only takes the repository name into account, i.e., "tensorflow-training", and not the actual image tag, i.e., "2.2.0-cpu/gpu-py37/...".

Therefore, I saw 3-4 attempts to run tests on all images simultaneously fail, because the name of the Training Job created by SageMaker was the same.

This random wait time between 0 and 60 seconds ensures that each image waits for a different amount of time before starting its test.

`Random(x="...")` is an instance of the `Random` class, where the unique combination of (image_uri + num_nodex) is given as a seed to ensure the time waited is always different.